pkg/virtualgateway/membership_designator.go

```diff
@@ -0,0 +1,103 @@
+package virtualgateway
+
+import (
+	"context"
+	appmesh "github.com/aws/aws-app-mesh-controller-for-k8s/apis/appmesh/v1beta2"
+	"github.com/aws/aws-app-mesh-controller-for-k8s/pkg/k8s"
+	"github.com/pkg/errors"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"strings"
+)
+
+// MembershipDesignator designates VirtualGateway membership for pods and namespaced AppMesh GatewayRoute CRs.
+type MembershipDesignator interface {
+	// DesignatePod will choose a VirtualGateway for given pod.
+	DesignatePod(ctx context.Context, pod *corev1.Pod) (*appmesh.VirtualGateway, error)
+	// DesignateNamespace will choose a VirtualGateway for given namespaced GatewayRoute CR.
+	DesignateNamespace(ctx context.Context, obj metav1.Object) (*appmesh.VirtualGateway, error)
```

Comment:
i'd favor like below to be more specific.
DesignateForGatewayRoute(ctx context.Context, obj *appmesh.GatewayRoute)
DesignateForPod(ctx context.Context, pod *corev1.Pod) (*appmesh.VirtualGateway, error)